src/app/static/src/app/components/forms/DynamicForm.vue

```diff
@@ -0,0 +1,101 @@
+<template>
+    <b-form :ref="id" class="dynamic-form" :validated="validated" novalidate>
+        <dynamic-form-control-section v-for="section in formMeta.controlSections"
+                                      :control-section="section">
+        </dynamic-form-control-section>
+        <button v-if="includeSubmitButton" class="btn btn-red" v-on:click="submit">{{submitText}}</button>
+    </b-form>
+</template>
+
+<script lang="ts">
+
+    import Vue from "vue";
+    import {BForm} from "bootstrap-vue";
+    import DynamicFormControlGroup from "./DynamicFormControlGroup.vue";
+    import DynamicFormControlSection from "./DynamicFormControlSection.vue";
+    import {DynamicForm} from "./types";
+    import {Dict} from "../../types";
+
+    interface Props {
+        formMeta: DynamicForm,
+        includeSubmitButton: boolean
+        submitText: string
+        id: string
+    }
+
+    interface SubmissionResult {
+        data: any
+        valid: boolean
+        missingValues: string[]
+    }
+
+    export default Vue.extend<{ validated: boolean }, { submit: (e: Event) => SubmissionResult }, { required: Dict<boolean> }, Props>({
+        name: "DynamicForm",
+        data() {
+            return {
+                validated: false
+            }
+        },
+        props: {
+            id: {
+                type: String,
+                default: Math.random().toString(36).slice(-5)
+            },
+            submitText: {
+                type: String,
+                default: "Submit"
+            },
+            includeSubmitButton: {
+                type: Boolean,
+                default: true
+            },
+            formMeta: {
+                type: Object
+            }
+        },
+        components: {
+            BForm,
+            DynamicFormControlGroup,
+            DynamicFormControlSection
+        },
+        computed: {
+            required() {
+                const requiredDict = {} as Dict<boolean>;
+                this.formMeta.controlSections.map(s => {
+                    s.controlGroups.map(g => {
+                        g.controls.map(c => {
+                            requiredDict[c.name] = c.required
+                        })
+                    })
+                });
+
+                return requiredDict;
+            }
+        },
+        methods: {
+            submit(e: Event) {
+                if (e) {
+                    e.preventDefault();
+                }
+                const form = this.$refs[this.id] as HTMLFormElement;
```

Comment:
I thought it would be nice to actually return the list of inputs which are missing, and I couldn't think of a better way of doing that with built in APIs?